subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -0,0 +1,497 @@
+		if (err) {
+			return err;
+		}
+#endif
+		initialized = true;
+	}
+
+	return 0;
+}
+
+int nrf_cloud_coap_connect(void)
+{
+	int err;
+
+	err = server_resolve();
+	if (err) {
+		LOG_ERR("Failed to resolve server name: %d", err);
+		return err;
+	}
+
+	LOG_DBG("Creating socket type IPPROTO_DTLS_1_2");
+	if (sock == -1) {
+		sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_DTLS_1_2);
+	}
+	LOG_DBG("sock = %d", sock);
+	if (sock < 0) {
+		LOG_ERR("Failed to create CoAP socket: %d.", -errno);
+		return -errno;
+	}
+
+	err = dtls_init(sock);
+	if (err < 0) {
+		LOG_ERR("Failed to initialize the DTLS client: %d", err);
+		goto fail;
+	}
+	authenticated = false;
+	if (cid_saved) {
+		err = dtls_session_load(sock);
+		cid_saved = false;
+		if (!err) {
+			LOG_INF("Loaded DTLS CID session");
+			authenticated = true;
+		} else if (err == -EAGAIN) {
+			LOG_INF("Failed to load DTLS CID session");
+		} else if (err == -EINVAL) {
+			LOG_INF("DLTS CID sessions not supported with current modem firmware");
+		} else {
+			LOG_ERR("Error on DTLS CID session load: %d", err);
+		}
+	} else {
+		LOG_INF("DTLS CID is not available");
+	}
+
+	/* Note: connect() of a SOCK_DGRAM socket does not actually connect, because
+	 * UDP is connectionless. The first actual traffic, including DTLS
+	 * handshake if needed, occurs when a sendto() or recvfrom() occurs on the
+	 * socket.
+	 */
+	err = connect(sock, (struct sockaddr *)&server,
+		      sizeof(struct sockaddr_in));
+	if (err < 0) {
+		LOG_ERR("Connect failed : %d", -errno);
+		goto fail;
+	}
+
+	err = nrf_cloud_coap_authenticate();
+	if (err < 0) {
+		goto fail;
+	}
+	return 0;
+
+fail:
+	(void)nrf_cloud_coap_disconnect();
+	return err;
+}
+
+static void auth_cb(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
+		    bool last_block, void *user_data)
+{
+	LOG_RESULT_CODE_INF("Authorization result_code:", result_code);
+	if (result_code < COAP_RESPONSE_CODE_BAD_REQUEST) {
+		authenticated = true;
+	}
+}
+
+static int nrf_cloud_coap_authenticate(void)
+{
+	int err;
+	char *jwt;
+
+	if (authenticated) {
+		LOG_INF("Already authenticated");
+		return 0;
+	}
+
+#if defined(CONFIG_MODEM_INFO)
+	char mfw_string[MODEM_INFO_FWVER_SIZE];
+	char ver_string[strlen(VER_STRING_FMT) +
+			MODEM_INFO_FWVER_SIZE +
+			strlen(BUILD_VERSION_STR) +
+			strlen(CDDL_VERSION)];
+
+	err = modem_info_get_fw_version(mfw_string, sizeof(mfw_string));
+	if (!err) {
+		err = snprintf(ver_string, sizeof(ver_string), VER_STRING_FMT,
+			       mfw_string, BUILD_VERSION_STR, CDDL_VERSION);
+		if ((err < 0) || (err >= sizeof(ver_string))) {
+			LOG_ERR("Could not format string");
+			return -ETXTBSY;
+		}
+	} else {
+		LOG_ERR("Unable to obtain the modem firmware version: %d", err);
+	}
+#else
+	char *ver_string = VER_STRING_FMT2;
+#endif
+
+	LOG_DBG("Generate JWT");
+	jwt = k_malloc(JWT_BUF_SZ);
+	if (!jwt) {
+		return -ENOMEM;
+	}
+	err = nrf_cloud_jwt_generate(NRF_CLOUD_JWT_VALID_TIME_S_MAX, jwt, JWT_BUF_SZ);
```

Comment:
Note that `sec_tag` is unsigned on nRF9160, so the format in `JWT_CMD_TEMPLATE_SEC_TAG` should use "%u" instead of "%d".

Reply:
OK.  That's relatively old code someone else wrote, but I can change it.  The `struct jwt_data` field `sec_tag` is defined as an `int` -- I will change that to `unsigned int` to match.